### DISCUSSION TITLE
how to load modules into a Compartment?

### DISCUSSION THREAD

**Discussion by dckc** [endojs/endo#1624] [Q&A] [answered]
In https://github.com/rchain/rchat/pull/9/files#r772716689 , @tgrospic reports that he tried...

```js
const r = c.evaluate(`
  import { config } from 'dotenv';
  config();
`);
```

but got: `SyntaxError#1: Cannot use import statement outside a module`.

This is somewhat unsurprising, but... then what? I suppose the answer is to use the compartment mapper somehow, but I'm puzzled as to exactly how.

Would you please
 - refine the examples in the README so that they are complete running examples, and/or
 - add a test showing how to do this sort of thing (without indirection through setup / scaffold stuff)

**Comment by ClemensLey** [comment]
Thank you for this amazing project!

I get the same error when I try to evaluate a module with an import. A running example or simple test would be super helpful.

**Comment by kriskowal** [comment, marked as answer]
@tgrecojs and @ClemensLey Please pardon the long wait for my answer!

The `Compartment` has an `import(specifier)` method, but for that to be useful, you need to create a `Compartment` with an appropriate `importHook` and `resolveHook`. But, those are very, very low-level and have no opinions about Node.js versus Web style import semantics. Regardless, that looks like:

- https://github.com/endojs/endo/blob/master/packages/ses/README.md#modules
- https://github.com/endojs/endo/blob/master/packages/ses/test/test-import.js

For import behavior that resembles Node.js, you would use `@endo/compartment-mapper`, which configures a `Compartment` for every package in an application, looking in all the places Node.js would look, and interpreting `package.json`. https://github.com/endojs/endo/blob/master/packages/compartment-mapper/README.md

But, even that is a little too low-level, because SES censors any program that looks like it might have an `import()` or `eval()` call in its sources. This is because the SES runtime needs to be fast, and so it is sloppy: it disqualifies a bunch of modules that are actually compatible. So, you need “evasive transforms”: a full parse of every module that replaces things that look like import() calls, but are only really in comments, and gets rid of HTML comments if they appear anywhere. We find that most libraries work at this level.

For evasive transforms, you need `@endo/bundle-source` and `@endo/import-bundle`. Our product at Agoric has not yet motivated us to create something like `@endo/import` that would do the evasive transforms and work directly on the host filesystem. For the moment, you have to create a bundle and then run it. But otherwise, this is the API that will get you where you need to be. 

- https://github.com/endojs/endo/blob/master/packages/bundle-source/README.md
- https://github.com/endojs/endo/blob/master/packages/import-bundle/README.md

We are working on a CLI that would make this all much more striahgtforward: `endo run script.js` would orchestrate all of the above and then call `main(powers)`. This is around the corner.

- Endo Pet Daemon Epic https://github.com/endojs/endo/issues/879

**Comment by dckc** [comment]
a simple example that doesn't try to emulate all of node.js would still be useful. For example, an `importHook` that just did

```
const moduleSource1 = '...';
return moduleSource1
```

**Comment by dckc** [comment]
ah... https://github.com/endojs/endo/blob/master/packages/ses/test/test-import.js seems to have the relevant stuff.
